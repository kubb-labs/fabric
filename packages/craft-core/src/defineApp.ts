import { write } from './fs.ts'
import { createFile, parseFile } from './parser.ts'
import type { Ref } from './reactive/ref.ts'
import type * as KubbFile from './types.ts'

const isFunction = (val: unknown): val is Function => typeof val === 'function'

type Component = any

type PluginInstallFunction<Options = any[]> = Options extends unknown[] ? (app: App, ...options: Options) => any : (app: App, options: Options) => any

export type ObjectPlugin<Options = any[]> = {
  install: PluginInstallFunction<Options>
}
export type FunctionPlugin<Options = any[]> = PluginInstallFunction<Options> & Partial<ObjectPlugin<Options>>

type Renderer = {
  run(): Promise<void> | void
  files: Ref<Array<KubbFile.File>>
  waitUntilExit(): Promise<void>
  output: Ref<string>
}

export type RootRenderFunction<HostElement = unknown> = (container: HostElement) => Renderer

export type Plugin<Options = any[], P extends unknown[] = Options extends unknown[] ? Options : [Options]> = FunctionPlugin<P> | ObjectPlugin<P>

export interface App<_HostElement = unknown> {
  _component: Component
  run(): Promise<void>
  use<Options>(plugin: Plugin<Options>, options: NoInfer<Options>): this
  write(): Promise<void>
  files: Array<KubbFile.File>
  waitUntilExit(): Promise<void>
  output: string
}

export type DefineApp<HostElement> = (rootComponent: Component) => App<HostElement>

export function defineApp<HostElement>(instance: RootRenderFunction<HostElement>): DefineApp<HostElement> {
  function createApp<HostElement>(rootComponent: Component): App<HostElement> {
    const installedPlugins = new WeakSet()
    const { run, waitUntilExit, files, output } = instance(rootComponent)

    const app: App<HostElement> = {
      _component: rootComponent,
      get files() {
        return files.value
      },
      get output() {
        return output.value
      },
      async run() {
        await run()

        await waitUntilExit()
      },
      waitUntilExit,
      async write() {
        for (const file of files.value) {
          const resolvedFile = createFile(file)
          const source = await parseFile(resolvedFile)

          await write(file.path, source)
        }
      },
      use(plugin: Plugin, ...options: any[]) {
        if (installedPlugins.has(plugin)) {
          console.warn('Plugin has already been applied to target app.')
        } else if (plugin && isFunction(plugin.install)) {
          installedPlugins.add(plugin)
          plugin.install(app, ...options)
        } else if (isFunction(plugin)) {
          installedPlugins.add(plugin)
          plugin(app, ...options)
        }

        return app
      },
    }

    return app
  }

  return createApp
}
