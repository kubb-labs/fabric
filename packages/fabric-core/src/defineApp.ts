import { FileManager } from './FileManager.ts'
import { isFunction } from 'remeda'
import type { Plugin } from './plugins/types.ts'
import type { Parser } from './parsers/types.ts'
import { AsyncEventEmitter } from './utils/AsyncEventEmitter.ts'
import type { AppContext, Component, AppEvents } from './App.ts'

import type { App } from './index.ts'

type RootRenderFunction<TApp extends App> = (app: TApp) => void | Promise<void>

export type DefineApp<TOptions> = (rootComponent?: Component, options?: TOptions) => App

export function defineApp<TOptions = unknown>(instance?: RootRenderFunction<App<TOptions>>): DefineApp<TOptions> {
  function createApp(options?: TOptions): App {
    const events = new AsyncEventEmitter<AppEvents>()
    const installedPlugins = new Set<Plugin>()
    const installedParsers = new Set<Parser>()
    const fileManager = new FileManager({ events })
    const context = {
      events,
      options,
      fileManager,
      installedPlugins,
      installedParsers,
    } as AppContext<TOptions>

    const app = {
      context,
      get files() {
        return fileManager.files
      },
      async addFile(...newFiles) {
        await fileManager.add(...newFiles)
      },
      async use(pluginOrParser, ...options) {
        const args = options

        if (pluginOrParser.type === 'plugin') {
          if (installedPlugins.has(pluginOrParser)) {
            console.warn('Plugin has already been applied to target app.')
          } else {
            installedPlugins.add(pluginOrParser)
          }

          if (pluginOrParser.inject && isFunction(pluginOrParser.inject)) {
            const injecter = pluginOrParser.inject

            const extraApp = (injecter as any)(app, ...args)
            Object.assign(app, extraApp)
          }
        }
        if (pluginOrParser.type === 'parser') {
          if (installedParsers.has(pluginOrParser)) {
            console.warn('Parser has already been applied to target app.')
          } else {
            installedParsers.add(pluginOrParser)
          }
        }

        if (pluginOrParser && isFunction(pluginOrParser.install)) {
          const installer = pluginOrParser.install

          await (installer as any)(app, ...args)
        }

        return app
      },
    } as App<TOptions>

    if (instance) {
      instance(app)
    }

    return app
  }

  return createApp
}
