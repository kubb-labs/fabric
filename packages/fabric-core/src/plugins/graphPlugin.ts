import { createPlugin } from './createPlugin.ts'
import type * as KubbFile from '../KubbFile.ts'
import { type Graph, TreeNode } from '../utils/TreeNode.ts'
import path from 'node:path'
import http from 'node:http'
import type { AddressInfo } from 'node:net'
import handler from 'serve-handler'
import { open } from '../utils/open.ts'

import { createFile } from '../createFile.ts'

type Options = {
  root: string
  /**
   * @default false
   */
  open?: boolean
}

type GetGraphOptions = {
  files: KubbFile.File[]
  root: string
}

export function getGraph({ files, root }: GetGraphOptions): Graph | undefined {
  const treeNode = TreeNode.fromFiles(files, root)

  if (!treeNode) {
    return undefined
  }

  return TreeNode.toGraph(treeNode)
}
const html = `
      <!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>File Graph</title>
    <script type="module">
      import { Network } from 'https://cdn.jsdelivr.net/npm/vis-network/standalone/esm/vis-network.min.js'

      async function main() {
        const res = await fetch('./graph.json')
        const { nodes, edges } = await res.json()
        const container = document.getElementById('graph')

        const network = new Network(
          container,
          { nodes, edges },
          {
            layout: { hierarchical: { direction: 'UD', sortMethod: 'directed' } },
            nodes: { shape: 'box', font: { face: 'monospace' } },
            edges: { arrows: 'to' },
            physics: false,
          },
        )
      }

      main()
    </script>
    <style>
      html, body, #graph { height: 100%; margin: 0; }
    </style>
  </head>
  <body>
    <div id="graph"></div>
  </body>
</html>
`

async function serve(root: string) {
  const server = http.createServer((req, res) => {
    return handler(req, res, {
      public: root,
      cleanUrls: true,
    })
  })

  server.listen(0, async () => {
    const { port } = server.address() as AddressInfo
    console.log(`Running on http://localhost:${port}/graph.html`)

    await open(`http://localhost:${port}/graph.html`)
  })
}

export const graphPlugin = createPlugin<Options>({
  name: 'graph',
  install(app, options) {
    if (!options) {
      throw new Error('Graph plugin requires options.root and options.mode')
    }

    app.context.events.on('write:start', async ({ files }) => {
      const root = options.root

      const graph = getGraph({ files, root })

      if (!graph) {
        return undefined
      }

      const graphFile = createFile({
        baseName: 'graph.json',
        path: path.join(root, 'graph.json'),
        sources: [
          {
            name: 'graph',
            value: JSON.stringify(graph, null, 2),
          },
        ],
      })

      const graphHtmlFile = createFile({
        baseName: 'graph.html',
        path: path.join(root, 'graph.html'),
        sources: [
          {
            name: 'graph',
            value: html,
          },
        ],
      })

      await app.context.fileManager.add(graphFile, graphHtmlFile)

      if (options.open) {
        await serve(root)
      }
    })
  },
})
