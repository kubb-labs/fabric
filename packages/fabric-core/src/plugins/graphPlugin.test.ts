import path from 'node:path'
import { beforeEach, describe, expect, test, vi } from 'vitest'

import * as GraphPluginModule from './graphPlugin.ts'
import { graphPlugin } from './graphPlugin.ts'
import { defineApp } from '../defineApp.ts'
import { createFile } from '../createFile.ts'
import type * as KubbFile from '../KubbFile.ts'

function makeFiles(count = 3): KubbFile.ResolvedFile[] {
  const files: KubbFile.ResolvedFile[] = [] as any
  for (let i = 0; i < count; i++) {
    files.push(
      createFile({
        path: path.join('src', `file${i}.ts`),
        baseName: `file${i}.ts`,
        sources: [
          {
            name: `S${i}`,
            value: `export const S${i} = ${i};`,
            isExportable: true,
            isIndexable: true,
          },
        ],
      }),
    )
  }
  return files
}

describe('graphPlugin', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
  })

  test('creates graph.json and graph.html on write:start', async () => {
    const app = defineApp()()
    await app.use(graphPlugin, { root: 'src', open: false })

    const files = makeFiles(2)

    await app.context.events.emit('write:start', { files })

    const graphJson = app.files.find((item) => item.baseName === 'graph.json')
    const graphHtml = app.files.find((item) => item.baseName === 'graph.html')

    expect(graphJson).toBeDefined()
    expect(graphJson?.sources).toMatchSnapshot()

    expect(graphHtml).toBeDefined()
    expect(graphHtml?.sources).toMatchSnapshot()
  })

  test('does nothing when getGraph returns undefined', async () => {
    const app = defineApp()()
    await app.use(graphPlugin, { root: 'out' })

    const files = makeFiles(1)

    const addSpy = vi.spyOn(app.context.fileManager, 'add')
    vi.spyOn(GraphPluginModule, 'getGraph').mockReturnValue(undefined)

    await app.context.events.emit('write:start', { files })

    expect(addSpy).not.toHaveBeenCalled()
  })
})
