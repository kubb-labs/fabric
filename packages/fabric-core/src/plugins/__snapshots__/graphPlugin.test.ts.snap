// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphPlugin > creates graph.json and graph.html on write:start 1`] = `
[
  {
    "name": "graph",
    "value": "{
  "nodes": [
    {
      "id": "src",
      "label": "src"
    },
    {
      "id": "src/file0.ts",
      "label": "file0.ts"
    },
    {
      "id": "src/file1.ts",
      "label": "file1.ts"
    }
  ],
  "edges": [
    {
      "from": "src",
      "to": "src/file0.ts"
    },
    {
      "from": "src",
      "to": "src/file1.ts"
    }
  ]
}",
  },
]
`;

exports[`graphPlugin > creates graph.json and graph.html on write:start 2`] = `
[
  {
    "name": "graph",
    "value": "
      <!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>File Graph</title>
    <script type="module">
      import { Network } from 'https://cdn.jsdelivr.net/npm/vis-network/standalone/esm/vis-network.min.js'

      async function main() {
        const res = await fetch('./graph.json')
        const { nodes, edges } = await res.json()
        const container = document.getElementById('graph')

        const network = new Network(
          container,
          { nodes, edges },
          {
            layout: { hierarchical: { direction: 'UD', sortMethod: 'directed' } },
            nodes: { shape: 'box', font: { face: 'monospace' } },
            edges: { arrows: 'to' },
            physics: false,
          },
        )
      }

      main()
    </script>
    <style>
      html, body, #graph { height: 100%; margin: 0; }
    </style>
  </head>
  <body>
    <div id="graph"></div>
  </body>
</html>
",
  },
]
`;
